File: frontend/src/types/machine.ts
export interface Machine {
  id: number;
  name: string;
  description?: string;
  machine_type?: string;
  manufacturer?: string;
  model?: string;
  serial_number?: string;
  status: string;
  ip_address?: string;
  location?: string;
  purchase_date?: string;
  purchase_price?: number;
  specs?: Record<string, unknown>;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface MachineCreate {
  name: string;
  description?: string;
  machine_type?: string;
  manufacturer?: string;
  model?: string;
  serial_number?: string;
  status?: string;
  ip_address?: string;
  location?: string;
  purchase_date?: string;
  purchase_price?: number;
  specs?: Record<string, unknown>;
  notes?: string;
}

export type MachineUpdate = Partial<MachineCreate>;

export interface MaintenanceTask {
  id: number;
  machine_id: number;
  title: string;
  description?: string;
  priority: string;
  status: string;
  scheduled_date?: string;
  completed_date?: string;
  recurrence_days?: number;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface MaintenanceTaskCreate {
  machine_id: number;
  title: string;
  description?: string;
  priority?: string;
  status?: string;
  scheduled_date?: string;
  recurrence_days?: number;
  notes?: string;
}

export interface MaintenanceTaskUpdate {
  title?: string;
  description?: string;
  priority?: string;
  status?: string;
  scheduled_date?: string;
  completed_date?: string;
  recurrence_days?: number;
  notes?: string;
}

export const MACHINE_TYPES = [
  "3D Printer",
  "CNC",
  "Laser Cutter",
  "Laser Engraver",
  "Pick & Place",
  "Reflow Oven",
  "Solder Paste Printer",
  "Soldering Station",
  "Hot Air Station",
  "Oscilloscope",
  "Multimeter",
  "Signal Generator",
  "Logic Analyzer",
  "Spectrum Analyzer",
  "Power Supply",
  "Electronic Load",
  "PCB Mill",
  "PCB Etcher",
  "UV Exposure Unit",
  "Thermal Camera",
  "Server",
  "Network Switch",
  "Router / Firewall",
  "UPS",
  "NAS",
  "Rack Cabinet",
  "KVM Switch",
  "Environmental Monitor",
  "Label Printer",
  "Microscope",
  "Fume Extractor",
  "Vacuum Pump",
  "Ultrasonic Cleaner",
  "Other",
] as const;

File: backend/app/schemas/machine.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict


class MachineBase(BaseModel):
    name: str
    machine_type: Optional[str] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    serial_number: Optional[str] = None
    ip_address: Optional[str] = None
    location: Optional[str] = None
    status: str = "offline"
    specs: Optional[dict[str, Any]] = None
    purchase_date: Optional[datetime] = None
    purchase_price: Optional[float] = None
    notes: Optional[str] = None
    pcb_origin_x: Optional[float] = None
    pcb_origin_y: Optional[float] = None
    nozzle_height_datum: Optional[float] = None
    default_mount_speed: Optional[int] = None


class MachineCreate(MachineBase):
    pass


class MachineUpdate(BaseModel):
    name: Optional[str] = None
    machine_type: Optional[str] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    serial_number: Optional[str] = None
    ip_address: Optional[str] = None
    location: Optional[str] = None
    status: Optional[str] = None
    specs: Optional[dict[str, Any]] = None
    purchase_date: Optional[datetime] = None
    purchase_price: Optional[float] = None
    notes: Optional[str] = None
    pcb_origin_x: Optional[float] = None
    pcb_origin_y: Optional[float] = None
    nozzle_height_datum: Optional[float] = None
    default_mount_speed: Optional[int] = None


class MachineResponse(MachineBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class MachineListResponse(BaseModel):
    items: list[MachineResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class MaintenanceTaskBase(BaseModel):
    title: str
    description: Optional[str] = None
    priority: str = "medium"
    status: str = "pending"
    scheduled_date: Optional[datetime] = None
    recurrence_days: Optional[int] = None
    notes: Optional[str] = None


class MaintenanceTaskCreate(MaintenanceTaskBase):
    pass


class MaintenanceTaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[str] = None
    status: Optional[str] = None
    scheduled_date: Optional[datetime] = None
    completed_date: Optional[datetime] = None
    recurrence_days: Optional[int] = None
    notes: Optional[str] = None


class MaintenanceTaskResponse(MaintenanceTaskBase):
    id: int
    machine_id: int
    completed_date: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class MachineCommand(BaseModel):
    command: str
    params: Optional[dict[str, Any]] = None
